File: ios/MochiShared/Input/ShiftState.swift
import Foundation

/// The three states shift can actually be in.
///
/// Modelled as an enum rather than a `Bool` + a separate `isCapsLocked` flag because the two-flag
/// version has an unrepresentable-but-reachable fourth state (locked *and* one-shot), and the bug
/// it produces — caps lock silently releasing after one letter — is exactly the kind of thing that
/// gets reported as "the keyboard is broken" rather than as a specific defect.
enum ShiftState: Equatable {
    /// Lowercase.
    case off
    /// Uppercase for exactly one character, then back to `off`. What the system does at the start
    /// of a sentence, and what a single shift tap gives you.
    case oneShot
    /// Uppercase until explicitly turned off. Reached by double-tapping shift.
    case locked

    var isUppercase: Bool { self != .off }

    /// Whether the shift key should render in its engaged style.
    var showsEngagedKey: Bool { self != .off }

    /// What a single tap on the shift key does from here.
    ///
    /// Tapping while locked releases the lock rather than dropping to one-shot — a user tapping a
    /// lit caps-lock key is asking for lowercase, not for one more capital.
    var afterShiftTap: ShiftState {
        switch self {
        case .off: return .oneShot
        case .oneShot: return .off
        case .locked: return .off
        }
    }

    /// Applied after a character is inserted.
    var afterCharacterInput: ShiftState {
        self == .oneShot ? .off : self
    }
}

File: android/app/src/main/java/com/mochi/keyboard/ime/render/ShiftState.kt
package com.mochi.keyboard.ime.render

/**
 * The three states shift can actually be in — ported from iOS's `ShiftState`. Modelled as an enum
 * rather than a boolean-plus-flag pair because the two-flag version has an unrepresentable-but-
 * reachable fourth state (locked *and* one-shot), whose symptom — caps lock silently releasing
 * after one letter — reads as "the keyboard is broken" rather than a specific defect.
 */
enum class ShiftState {
    /** Lowercase. */
    OFF,
    /** Uppercase for exactly one character, then back to OFF. */
    ONE_SHOT,
    /** Uppercase until explicitly turned off. */
    LOCKED;

    val isUppercase: Boolean get() = this != OFF
    val showsEngagedKey: Boolean get() = this != OFF

    /** What a single tap on the shift key does from here. Tapping while locked releases the lock
     * rather than dropping to one-shot — a lit caps-lock key being tapped is a request for
     * lowercase, not for one more capital. */
    fun afterShiftTap(): ShiftState = when (this) {
        OFF -> ONE_SHOT
        ONE_SHOT -> OFF
        LOCKED -> OFF
    }

    fun afterCharacterInput(): ShiftState = if (this == ONE_SHOT) OFF else this
}
